game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTransportUtils.java

```diff
@@ -121,6 +127,29 @@ public static int findMaxMovementForTransports(final List<ProPurchaseOption> sea
         }
       }
     }
+
+    // If extra space try to replace last unit with stronger unit
+    if (!selectedUnits.isEmpty() && capacityCount < capacity) {
+      final Unit lastUnit = selectedUnits.get(selectedUnits.size() - 1);
+      final int lastUnitCost = UnitAttachment.get(lastUnit.getType()).getTransportCost();
+      units.removeAll(selectedUnits);
+      Comparator<Unit> comparator;
+      if (Matches.unitIsLandTransport().test(transport)) {
+        comparator = Comparator.<Unit>comparingInt(u -> TripleAUnit.get(u).getMovementLeft())
+            .thenComparing(getDecreasingAttackComparator(transport.getOwner()));
+      } else {
+        comparator = getDecreasingAttackComparator(transport.getOwner());
+      }
+      units.sort(comparator);
+      for (final Unit unit : units) {
+        final int cost = UnitAttachment.get(unit.getType()).getTransportCost();
+        if (comparator.compare(unit, lastUnit) < 0 && (capacityCount - lastUnitCost + cost <= capacity)) {
+          selectedUnits.remove(lastUnit);
+          selectedUnits.add(unit);
+          break;
```

Comment:
I'm not quite getting the logic here.
If the comparator is being used here manually, why does the list even need to be sorted, or if it needs to be sorted, why can't we rely on the order in the list (I.e. Checking for unit equals lastUnit and if so, abort the loop because all remaining elements have to be greater.

Reply:
So the already selected units are removed from the `units` list and the last unit that was selected becomes the unit we are trying to replace. The list is re-sorted to ignore transport cost and just find the strongest units to try to replace it with. The comparator is used manually to check that the replacing unit from the list is actually stronger than the already selected unit. 

An example of what this code does on say revised/global is it first loads a transport (capacity 5) with 2 infantry (each cost 2) then it tries to fill the extra space of 1 (5 - 4) by replacing an infantry with an artillery/tank (cost 3). The code first tries to pack in as many units as possible into the transport then tries to replace the last unit with a stronger unit if possible.

Reply:
Moved the comparator check to the first portion of loop. As since the list is sorted, if you get to a unit that isn't stronger than the `lastUnit` then you don't need to check the rest.